### PR TITLE
Allow unordered KID:key by matching KIDs to streams

### PR DESCRIPTION
**Pull request by its-wizza**
Not extensively tested

### CODE REVIEW

File: dashdrm/dashdrm.py
Comment (titus-au) [inline dashdrm/dashdrm.py:1]: I've sent you emails with comments...

File: dashdrm/dashdrm.py [hunk 1]
```diff
@@ -830,6 +847,137 @@ def sortby_bandwidth(dash_stream: DASHStreamDRM) -> float:
...
+            for _ in readers:
+                out_keys.append(keys[key])
+                key += 1
+                # If we had more streams than keys, start with the first audio key again
+                if key == len(keys):
+                    key = 1
+            return out_keys
+
+        if not kid_lookup:
+            if len(in_keys) > 1:
```
Comment (titus-au) [inline dashdrm/dashdrm.py:886]: Should this be >= 1?

File: dashdrm/dashdrm.py [hunk 2]
```diff
@@ -147,35 +149,58 @@ def _get_streams(self):
...
                 try:
-                    int(key[-1], 16)
+                    int(key, 16)
                 except ValueError as err:
                     raise FatalPluginError("Expecting 128bit key in 32 hex digits, but the key contains invalid hex.")
             elif key_len != 32:
                 raise FatalPluginError("Expecting 128bit key in 32 hex digits.")
-            return_keys.append(key[-1])
+            return_keys.append((kid, key))
+        self.session.options["store-representation-kid"] = has_kid
```
Comment (titus-au) [inline dashdrm/dashdrm.py:203]: I don't quite get the reason for this. It looks like this is used later to avoid re-testing whether we have kids with keys. However, as it stands, it only indicates whether we have at least 1 kid with a key, when for kid matching to work we need kids for every key. If we have kid for every key, then all we need to do later is test for the first kid not being None, and there's no need to store this? Am I missing something?

Comment (its-wizza) [reply dashdrm/dashdrm.py:203]: KID matching can work when only one key with a KID is provided, if that is the only KID present across multiple streams in the manifest. 

Consider if multiple keys are provided for a manifest with encrypted video, only one key has a KID, and it is not the first key provided via the decryption-key argument. With the current implementation, the key will be correctly identified via KID matching and be provided for the streams requiring it.

If we ignore KIDs if at least one is not provided, then the first key will be used to attempt decrypting the video which will not work.

Of course it is an edge case and not realistic but I think better to attempt KID matching where possible.

Hopefully I made sense here

File: dashdrm/dashdrm.py (hunk 1 above)
Comment (its-wizza) [reply dashdrm/dashdrm.py:886]: I only log a message if more than one key is provided. If only one is provided, it is obvious that it will be applied across all encrypted streams as necessary. Whereas if two keys are provided, exact behaviour depends on the number of streams so I thought it was worth including a debug log to indicate the key matching method. Does that make sense? Maybe that is overthinking it since it's a debug log anyway.
Open to changes with the logging. Not sure how effective it is.

File: dashdrm/dashdrm.py (hunk 2 above)
Comment (titus-au) [reply dashdrm/dashdrm.py:203]: Yeah no worries. I typed this last night before I thought it through.

File: dashdrm/dashdrm.py (hunk 1 above)
Comment (titus-au) [reply dashdrm/dashdrm.py:886]: Yeah that's a bit obscure I think. If the kid_lookup dict is empty, we'll be reverting to positional keys so that's worth logging I think.
